peer.go

```diff
@@ -454,6 +424,128 @@ func (p *peer) loadActiveChannels(chans []*channeldb.OpenChannel) error {
+		Peer:                  p,
+		DecodeHopIterators:    p.server.sphinx.DecodeHopIterators,
+		ExtractErrorEncrypter: p.server.sphinx.ExtractErrorEncrypter,
+		FetchLastChannelUpdate: fetchLastChanUpdate(
+			p.server, p.PubKey(),
+		),
+		DebugHTLC:      cfg.DebugHTLC,
+		HodlMask:       cfg.Hodl.Mask(),
+		Registry:       p.server.invoices,
+		Switch:         p.server.htlcSwitch,
+		Circuits:       p.server.htlcSwitch.CircuitModifier(),
+		ForwardPackets: p.server.htlcSwitch.ForwardPackets,
+		FwrdingPolicy:  *forwardingPolicy,
+		FeeEstimator:   p.server.cc.feeEstimator,
+		BlockEpochs:    blockEpoch,
+		PreimageCache:  p.server.witnessBeacon,
+		ChainEvents:    chainEvents,
+		UpdateContractSignals: func(signals *contractcourt.ContractSignals) error {
+			return p.server.chainArb.UpdateContractSignals(
+				*chanPoint, signals,
+			)
+		},
+		OnChannelFailure: func(chanID lnwire.ChannelID,
+			shortChanID lnwire.ShortChannelID,
+			linkErr htlcswitch.LinkFailureError) {
+
+			// The link has notified us about a failure. We launch
+			// a go routine to stop the link, disconnect the peer
+			// and optionally force close the channel. We must
+			// launch a goroutine since we must let
+			// OnChannelFailure return in order for the link to
+			// completely stop in the call to RemoveLink.
+			p.wg.Add(1)
+			go func() {
+				defer p.wg.Done()
+
+				// We begin by removing the link from the
+				// switch, such that it won't be attempted used
+				// for any more updates.
+				err := p.server.htlcSwitch.RemoveLink(chanID)
+				if err != nil {
+					peerLog.Errorf("unable to stop "+
+						"link(%v): %v", shortChanID,
+						err)
+				}
+
+				// If the error encountered was severe enough,
+				// we'll now force close the channel.
+				if linkErr.ForceClose {
+					peerLog.Warnf("Force closing link(%v)",
+						shortChanID)
+
+					closeTx, err := p.server.chainArb.ForceCloseContract(*chanPoint)
+					if err != nil {
+						peerLog.Errorf("unable to "+
+							"force close link(%v):"+
+							" %v", shortChanID, err)
+					} else {
+						peerLog.Infof("channel(%v) "+
+							"force closed with "+
+							"txid %v", shortChanID,
+							closeTx.TxHash())
+					}
+				}
+
+				// Send an error to the peer, why we failed the
+				// channel.
+				if linkErr.ShouldSendToPeer() {
+					// If SendData is set, send it to the
+					// peer. If not, we'll use the standard
+					// error messages in the payload. We
+					// only include sendData in the cases
+					// where the error data does not
+					// contain sensitive information.
+					data := []byte(linkErr.Error())
+					if linkErr.SendData != nil {
+						data = linkErr.SendData
+					}
+					err := p.SendMessage(&lnwire.Error{
+						ChanID: chanID,
+						Data:   data,
+					}, true)
+					if err != nil {
+						peerLog.Errorf("unable to "+
+							"send msg to remote "+
+							"peer: %v", err)
+					}
+				}
+
+				// Initiate disconnection.
+				p.Disconnect(linkErr)
```

Comment:
Thinking about this disconnect, what if we have other channels that didn't fail? Is just removing the link sufficient?

Reply:
FWIW the existing behavior would _always_ remove the connection regardless of if there were other existing channels w/ that peer.

Reply:
Added TODO, in favour of keeping the existing behaviour for now.